text/0225-configuration.md

```diff
@@ -0,0 +1,410 @@
+          schedule_delay: 5000
+          # Sets the maximum allowed time to export data.
+          #
+          # Environment variable: OTEL_BSP_EXPORT_TIMEOUT
+          export_timeout: 30000
+          # Sets the maximum queue size.
+          #
+          # Environment variable: OTEL_BSP_MAX_QUEUE_SIZE
+          max_queue_size: 2048
+          # Sets the maximum batch size.
+          #
+          # Environment variable: OTEL_BSP_MAX_EXPORT_BATCH_SIZE
+          max_export_batch_size: 512
+          # Sets the exporter. Exporter must refer to a key in sdk.tracer_provider.exporters.
+          #
+          # Environment variable: OTEL_TRACES_EXPORTER
+          exporter: zipkin
+  # Configure the meter provider.
+  ...
+```
+
+### Environment variable substitution
+
+Configuration files *MUST* support environment variable expansion. While this accommodates the scenario in which a configuration file needs to reference sensitive data and is not able to be stored securely, environment variable expansion is not limited to sensitive data.
+
+The syntax for environment variable expansion mirrors the [collector](https://opentelemetry.io/docs/collector/configuration/#configuration-environment-variables).
+
+For example, given an environment where `API_KEY=1234`, the configuration file contents:
+
+```yaml
+scheme_version: 0.1
+sdk:
+  tracer_provider:
+    exporters:
+      otlp:
+        endpoint: https://example.host:4317/v1/traces
+        headers:
+          api-key: ${env:API_KEY}
+```
+
+Result in the following after substitution:
+
+```yaml
+scheme_version: 0.1
+sdk:
+  tracer_provider:
+    exporters:
+      otlp:
+        endpoint: https://example.host:4317/v1/traces
+        headers:
+          api-key: 1234
+```
+
+Implementations *MUST* perform environment variable substitution before validating and parsing configuration file contents.
+
+### Version guarantees & backwards compatibility
+
+Each version of the configuration schema carries a major and minor version. Configurations specify the major and minor version they adhere to. Before reaching 1.0, each minor version change is equivalent to major version change. That is, there are no guarantees about compatibility and all changes are permitted. As of 1.0, we provide the following stability guarantees:
+
+* For major version: No guarantees.
+* For minor versions:
+  * Property keys will not change. Although the addition of optional properties is permitted and the removal of properties is permitted, this *MUST* not be abused to rename property keys.
+  * Property value types will not change, except that integers MAY become floating points.
+  * No additional required properties.
+
+Allowable changes:
+
+* For major versions: All changes are permitted.
+* For minor versions:
+  * Addition of optional properties.
+  * Required property keys may become optional.
+  * Removal of properties, provided that the property key is not reused in the future.
```

Comment:
What's the rationale for allowing removal? It seems pretty disruptive to me.

Reply:
No strong rational. Generally was trying to mirror schema evolution rules of other systems like [protobuf](https://protobuf.dev/programming-guides/proto3/#updating) and [avro](https://docs.confluent.io/platform/current/schema-registry/avro.html), which both allow removal of fields.

This is far from an exhaustive list of the types of changes that are possible in json schema. There are [things like](https://json-schema.org/understanding-json-schema/) conditions, composition, regular expression, etc which aren't obvious in terms of backwards compatibility. I think we'll have to have general goals about not breaking users, and build up a set of rules as we encounter new proposals for changes.

Reply:
These would mean that the "validation/parsing/loading" would not be "strict", right? So if one makes e.g. `instrumetations:` instead of ``instrumentations:` it will be a NOOP? I do not think it would be desired.

Motivating example: https://www.jsonschemavalidator.net/s/jIlDubSn

I think we could only deprecate a field in a minor version.

```suggestion
  * Deprecation of properties.
```

Reply:
We discussed this in the 3/6 Configuration Working Group and decided that its premature to specify the allowable changes in the minor versions.

We think its best to:
* Keep the language which says "Before reaching 1.0, each minor version change is equivalent to major version change. That is, there are no guarantees about compatibility and all changes are permitted."
* Adjust the compatibility guarantees and allowable changes for minor version to simply be "TBD". We think this better reflects the reality that we'll need to develop the evolution rules over time as we encounter new situations.

Proposed text change [here](https://github.com/open-telemetry/oteps/pull/225/files#r1128610084).